posts/inside-rust/2021-05-15-What-the-error-handling-project-group-is-working-towards.md

```diff
@@ -0,0 +1,235 @@
+
+## Duplicate Information Issue
+
+With these fixes in place it will become easy to chain errors and report them completely and consistently. However there is a hazard that `Error` implementors need to be aware of in this system: duplicate information.
+
+Imagine an error like the one in the previous example, except instead of each error printing its own message and returning the next error via `source`, they also include their source's error message after their own. That way when we print the outer error's `Display` output we see all of the error messages, not just the first in the chain.
+
+```rust
+println!("Error: {}", error);
+
+// Outputs:
+// Error: outermost error: second error: root error
+```
+
+Now, what happens we then print this same error type with `Report` expecting that we need to iterate over the sources and print them too?
+
+```rust
+println!("Error: {:#}", error.report());
+
+// Outputs
+// Error: outermost error: second error: root error
+//
+// Caused by:
+//    0: second error: root error
+//    1: root error
+```
+
+The source error messages get duplicated! With the multi-line output of `anyhow` and `eyre` we get this nice little triangle shape to the error report, which you've probably encountered if you've ever used these libraries before. We can no longer separate the error messages of the individual errors in the chain of errors because this error type concatenates the sources manually and returns them via the `source` function. This also restricts how we can format our error reports. If we want a consistent report format and we have a dependency that concatenates errors in a single line we are forced to do so as well ourselves throughout our entire application. If, on the other hand, we have two dependencies that concatenate errors in different ways, well, we're out of luck.
+
+So how do we avoid this? We adopt a consistent separation for `Display` and `source` implementations.
+
+## Guidelines for implementing `Display::fmt` and `Error::source`
+
+To resolve this issue, project error handling recently created a guideline for [how to implement `Display::fmt` and `Error::source`](https://github.com/rust-lang/project-error-handling/issues/27#issuecomment-763950178). In it we make the following recommendation:
+
+**An error type with a source error should either return that error via `source` or include that source's error message in it's own `Display` output, but never both.**
+
+We figure the default will be to return error's via source, so that source errors can be reacted to via `downcast` when appropriate. In coming up with this recommendation we had to figure out what the `Error` trait's primary role is in Rust. After discussing it with the library team we concluded that reporting should be treated as the primary role, and that reacting via `downcast` should come second when designing error types.
```

Comment:
```suggestion
We figure the default will be to return errors via source, so that source errors can be reacted to via `downcast` when appropriate. In coming up with this recommendation we had to figure out what the `Error` trait's primary role is in Rust. After discussing it with the library team we concluded that reporting should be treated as the primary role, and that reacting via `downcast` should come second when designing error types.
```